Fix invalid malloc failures in `PEM_write_bio_PKCS8PrivateKey*()`

RaisinTen opened the pull request:
When `PEM_write_bio_PKCS8PrivateKey()` and `PEM_write_bio_PKCS8PrivateKey_nid()`
were passed empty passphrase strings, `OPENSSL_memdup()` was incorrectly getting
used for 0 bytes size allocations, which resulted in malloc failures.

Fixes: https://github.com/openssl/openssl/issues/17506

Signed-off-by: Darshan Sen <raisinten@gmail.com>

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

mattcaswell commented on crypto/o_str.c:
```diff
@@ -53,7 +54,7 @@ void *CRYPTO_memdup(const void *data, size_t siz, const char* file, int line)
 
     ret = CRYPTO_malloc(siz, file, line);
-    if (ret == NULL) {
+    if (ret == NULL && errno == ENOMEM) {
```
Calling memcpy with a NULL value for `ret` is (I think) undefined behaviour. So this change does not look correct to me. Possibly it is better to additionally confirm that `siz` is > 0 in the line `if (data == NULL || siz >= INT_MAX)` above. Regardless of whether `malloc` has failed with ENOMEM, `CRYPTO_malloc` has failed if it gives a NULL response and it seems to me that an `ERR_R_MALLOC_FAILURE` error is appropriate.

RaisinTen replied:
@mattcaswell Thanks for the suggestion but I removed the change in `CRYPTO_memdup()` because https://github.com/openssl/openssl/pull/17507#issuecomment-1013064386 suggested that the change should be made from the caller's side.

bernd-edlinger commented on crypto/passphrase.c:
```diff
@@ -42,7 +43,7 @@ int ossl_pw_set_passphrase(struct ossl_passphrase_data_st *data,
     data->_.expl_passphrase.passphrase_copy =
         OPENSSL_memdup(passphrase, passphrase_len);
-    if (data->_.expl_passphrase.passphrase_copy == NULL) {
+    if (data->_.expl_passphrase.passphrase_copy == NULL && errno == ENOMEM) {
```
This feels wrong, because it depends on wether OPENSSL_malloc(0) returns NULL or not,
but that is implementation dependent, and later we have this:
https://github.com/bernd-edlinger/openssl/blob/71396cd048072b69559b46d98cfebfd4474cd712/crypto/passphrase.c#L232-L238
so a NULL will not do the right thing.

bernd-edlinger replied:
I think it would be safer to change the statement above to
```C
data->_.expl_passphrase.passphrase_copy =
        passphrase_len != 0 ? OPENSSL_memdup(passphrase, passphrase_len)
                                          : OPENSSL_malloc(1);
```

bernd-edlinger replied:
Of course this PR needs a test case.

RaisinTen replied:
@bernd-edlinger thanks for the suggestion, I incorporated it into the PR and it seems to fix the issue mostly. I have also added some tests but I'm not very sure why I'm still hitting this failure only for the last test where I'm passing a callback:
```js
[Error: error:1400006B:UI routines::processing error] {
  opensslErrorStack: [
    'error:1C80009F:Provider routines::unable to get passphrase',
    'error:07880028:common libcrypto routines::UI lib'
  ],
  library: 'UI routines',
  reason: 'processing error',
  code: 'ERR_OSSL_UI_PROCESSING_ERROR'
}
```
I'm guessing it's an incorrect invocation? PTAL.

RaisinTen replied:
resolving as the added test passes CI

t8m commented:
In general this is wrong place to fix your problem. This needs to be properly handled at the callers to not try 0 bytes size allocation or memory duplication.

RaisinTen commented:
> In general this is wrong place to fix your problem. This needs to be properly handled at the callers to not try 0 bytes size allocation or memory duplication.

@t8m thanks, updated accordingly.

t8m reviewed (dismissed):
LGTM. To be able to accept the PR we will need a CLA signed from you. Please follow the process described at https://www.openssl.org/policies/cla.html

t8m commented:
The CI failures are relevant.

RaisinTen commented:
@t8m it's just the callback variation that fails which was happening even before this change was added. I'm suspecting that's happening due to an incorrect invocation of the API? (suggestions would be great!) I've commented out that test for now with a TODO. If CI passes now (not sure why these tests aren't run for me locally, will investigate later) and y'all are good with the change, I'll try to sign the CLA.

t8m commented:
Please re-enable the last test case. The error can be fixed by changing https://github.com/openssl/openssl/blob/master/crypto/ui/ui_util.c#L117 to `if (len < 0)`

t8m commented:
Also please sign and send the CLA.

bernd-edlinger commented:
> Please re-enable the last test case. The error can be fixed by changing https://github.com/openssl/openssl/blob/master/crypto/ui/ui_util.c#L117 to `if (len < 0)`

but please be careful, return values 0 and -1 have a special meaning,
and the user callback has no way to return 0 any more with this change.
this is the caller:
https://github.com/openssl/openssl/blob/5288303da96084b41b062d99eb37177fb4cf471e/crypto/ui/ui_lib.c#L516-L530
and this is the callee:
https://github.com/openssl/openssl/blob/5288303da96084b41b062d99eb37177fb4cf471e/crypto/ui/ui_util.c#L111-L121

t8m commented:
@bernd-edlinger the callback is the pem callback which returns 0 if the password is zero length, not that the callback has failed. We should not return 0 from ui_read() if the pem callback did not fail but simply returned 0 length password.
The ui_read() will return 0 if UI_set_result_ex() fails, and that is correct.
So I am sure that the change I propose is correct.

bernd-edlinger commented:
Well, I just said one should be very careful, when changing
the meaning of the return value of any user callback.

I am also nervous about this:
https://github.com/openssl/openssl/blob/f59d72f027da90edcccad5cc78c94d3099fadecf/crypto/pem/pem_pk8.c#L138-L142

It would be better to change this too, and more importantly also have
test coverage for this code path.

But, what worries me, is when I look at the function `i2d_PKCS8PrivateKey_bio`
this made me think that currently a user callback can return an error by
returning 0 or -1 here.  Likewise the pem_callback currently can return 0
to raise a hard error, while return -1 means something like user abort.
So there is a subtle difference in the error handling.
Probably we loose a way to signal a hard error from the user callback.

t8m commented:
> Well, I just said one should be very careful, when changing the meaning of the return value of any user callback.

But we are not changing it! We are correcting the mistake at the callers.

If you look at the implementation of PEM_def_callback which is the default implementation of this pem_password_cb() and if you look at the example of the pem_password_cb implementation at the manual page you can clearly see that returning 0 means zero length password and not and error! Citation from the manual page: `The callback must return the number of characters in the passphrase or -1 if an error occurred.`

t8m commented:
> I am also nervous about this:

This is just another bug - the condition should be `if (klen < 0)` if we want/can support 0 length passwords there.

t8m commented:
> But, what worries me, is when I look at the function `i2d_PKCS8PrivateKey_bio`

And that's another place where 0 byte passwords are unsupported. Which does not necessarily mean they cannot be. I can only repeat the manual page says fairly explicitly that just -1 is an error.

RaisinTen commented:
@t8m I changed the `<=` to `<` like you asked me to, let's see if CI passes this time! Is there any chance I can sign the CLA without taking a printout of https://www.openssl.org/policies/openssl_icla.pdf? May I add a picture of my signature instead like how docusign works?

bernd-edlinger commented:
Can you please also look at changing this one:
https://github.com/openssl/openssl/blob/f59d72f027da90edcccad5cc78c94d3099fadecf/crypto/pem/pem_pk8.c#L138-L142

and can someone figure out a way how that can be executed?

bernd-edlinger commented on test/evp_pkey_provided_test.c:
```diff
@@ -128,6 +128,11 @@ static int compare_with_file(const char *alg, int type, BIO *membio)
 
+static int pass_cb(char *buf, int size, int rwflag, void *u)
+{
+    return 0;
```
and I wonder if a negative test, where the callback return -1 would make sense too.

t8m replied:
Yes, that would make sense to add.

RaisinTen replied:
@bernd-edlinger added the test

bernd-edlinger replied:
excellent!

t8m commented:
> and can someone figure out a way how that can be executed?

That would be executed for a legacy pkey I suppose.

t8m commented:
> May I add a picture of my signature instead like how docusign works?

Yes, that is acceptable.

RaisinTen commented:
> Can you please also look at changing this one:
> 
> https://github.com/openssl/openssl/blob/f59d72f027da90edcccad5cc78c94d3099fadecf/crypto/pem/pem_pk8.c#L138-L142
> 
> and can someone figure out a way how that can be executed?

This can only de done through `PEM_write_bio_PKCS8PrivateKey_nid()`. Added a second commit to apply the fix you mentioned and added tests too but it fails because `p8` is assigned a null value here:
https://github.com/openssl/openssl/blob/e1cd94f2dca4056ce042c62b89c468dffc088033/crypto/pem/pem_pk8.c#L146
Not sure where I should make a change to fix this. 🤔

RaisinTen commented:
> Yes, that is acceptable.
@t8m thanks, signed the CLA :)

bernd-edlinger commented:
For your new test, You have to use one of the NIDs listed here:
https://github.com/openssl/openssl/blob/e1cd94f2dca4056ce042c62b89c468dffc088033/crypto/evp/evp_pbe.c#L36-L65

For instance `NID_pbe_WithSHA1And3_Key_TripleDES_CBC` should work.

RaisinTen commented:
@bernd-edlinger thanks, that fixed the test failure, PTAL :)

bernd-edlinger commented on test/evp_pkey_provided_test.c:
```diff
@@ -140,6 +150,33 @@ static int test_print_key_using_pem(const char *alg, const EVP_PKEY *pk)
+                                                     NULL, 0, pass_cb_error,
+                                                     NULL))
+        || !TEST_true(PEM_write_bio_PKCS8PrivateKey_nid(
+            bio_out, pk, NID_pbe_WithSHA1And3_Key_TripleDES_CBC,
```
Most other places where NID_pbe_WithSHA1And3_Key_TripleDES_CBC are used,
are guarded with `#ifndef OPENSSL_NO_DES`.  I guess it would be better to try that
only if the DES cipher is available

RaisinTen replied:
@bernd-edlinger thanks, added the guard, PTAL

bernd-edlinger reviewed (changes requested):
ci failure is relevant

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to master and 3.0 branches (resolving the CHANGES.md conflict manually). Thank you for your contribution.

RaisinTen commented:
@t8m thanks! Should this PR be closed, now that my commits have landed on the relevant branches or is there anything else I can help out with?
